### ISSUE TITLE
module avail enters an apparently infinite loop

### ISSUE THREAD

**Issue by damianam** [TACC/Lmod#227]
An user of ours has stumbled upon this issue. Steps to reproduce it:

-Have a system cache that contains `/some/path`
-Have an user with a module like this in `$HOME/brokenmodule`:

```
$ cat $HOME/brokenmodule/test
#%Module1.0#####################################################################
module use /some/path
```

-Type `module use $HOME/brokenmodule`
-Type `module avail`

It seems like lmod tries to rebuild the cache and enters an infinite loop while doing it. If `/some/path` is not in the system cache it works as expected.

**Comment by rtmclay** [comment]
Which version of Lmod?

**Comment by damianam** [comment]
In production we have `6.4.2`, but I've tested it with `HEAD` too.

`module --config` is actually broken for me in the latest release. I haven't had the time to figure out why:

```
/usr/bin/lua: /homeb/zam/alvarez/sw/lmod/lmod/libexec/../tools/BeautifulTbl.lua:135: bad argument #1 to 'length' (string expected, got boolean)
stack traceback:
	[C]: in function 'length'
	/homeb/zam/alvarez/sw/lmod/lmod/libexec/../tools/BeautifulTbl.lua:135: in function '_build_tbl'
	/homeb/zam/alvarez/sw/lmod/lmod/libexec/../tools/BeautifulTbl.lua:102: in function 'new'
	/homeb/zam/alvarez/sw/lmod/lmod/libexec/Configuration.lua:292: in function 'report'
	/homeb/zam/alvarez/sw/lmod/lmod/libexec/lmod:402: in function 'main'
	/homeb/zam/alvarez/sw/lmod/lmod/libexec/lmod:514: in main chunk
	[C]: ?
```

In `6.4.2` the output is:

```
Modules based on Lua: Version 6.4.2  2016-06-16 06:28
    by Robert McLay mclay@tacc.utexas.edu

Description                      Value
-----------                      -----
Allow TCL modulefiles            yes
Auto swapping                    yes
Case Independent Sorting         no
Colorize Lmod                    yes
Disable Same Name AutoSwap       no
Using dotfiles                   yes
Allow duplicate paths            yes
Require Exact Match/no defaults  no
Export the module command        YES
Lmod version                     6.4.2
Lua Version                      Lua 5.1
System lua_json
System lua-term
Active lua-term                  true
MODULERCFILE                     /usr/local/software/lmod/etc/rc
avail: Include modulepath dir    no
MODULEPATH_ROOT                  /usr/local/software/modulefiles
number of cache dirs             1
Version ordering                 modern
Pager                            /usr/bin/less
Pager Options                    -XqMREF
Path to HashSum                  /usr/bin/sha1sum
Path to Lua                      /usr/bin
Pin Versions in restore          no
Pkg Class name                   Pkg
Lmod prefix                      /usr/local/software
Prepend order                    normal
Redirect to stdout               no
Supporting Full Settarg Use      no
Site Pkg location                standard
Ignore Cache                     no
Cached loads                     yes
User cache valid time(sec)       86400
Write cache after (sec)          2
Threshold (sec)                  1
Tmod prepend PATH Rule           no
uname -a                         Linux jrl02 3.10.0-327.36.3.el7.x86_64 #1 SMP Mon Oct 24 16:09:20 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
Admin file                       /usr/local/software/lmod/etc/admin.list
Does Admin file exist            no


Active RC file(s):
------------------
/usr/local/software/lmod/6.4.2/init/lmodrc.lua
/usr/local/software/FZJ/lmodrc.lua


Cache Directory                                 Time Stamp File
---------------                                 ---------------
/usr/local/software/jureca/moduleData/cacheDir  /usr/local/software/jureca/moduleData/cacheDir/../system.txt
```

**Comment by rtmclay** [comment]
This endless loop problem has always existed in Lmod before Lmod 7.  It cannot be avoided in the way that spider works for previous versions of Lmod. This is the same problem that happens with the use.own module.

Now if you can show this problem in Lmod 7, I'll try to fix it. However, I am unable to reproduce it in Lmod 7.  It works fine for me and I followed the instructions you gave me. 

I have pushed to head my test and a fix for the problem you reported why ml --config didn't work.  I believe that ml --config problem is because the cache file location and timestamp file are not defined well in your lmodrc.lua file.  This fix to HEAD should protect Lmod from that kind of problem.

Please try using the latest HEAD to see if you can run "ml --config" and see if you can reproduce the module avail problem.  If you can reproduce then run the following:
```
    $ module -D -D avail 2> avail.log
```
and email me about 20 MBytes worth of avail.log.  Thanks!

**Comment by damianam** [comment]
Sorry it took me that long to respond.

I've tried again with both `HEAD` and `6.4.2`. What I thought it was triggering the issue, was that lmod started rebuilding the cache. As the user got into an infinite loop doing that, I thought I was getting into it too, since the rebuilding of the cache was somehow unexpected. So I killed lmod at that point. I was more patience now, and it eventually finishes without any major fuss. Bottom line, I couldn't reproduce the issue, I just mistakenly thought I was doing it.

I'll close the issue now, since maybe the problem lied somewhere else that we can't pinpoint right now and it doesn't hinder anybody's work. Sorry for the noise!
